File: backend/src/services/features/authenticationService.ts
import type { PrismaClient, User } from "#/prisma/client";
import type { IServiceContainer } from "#/dependencyInjection";
import { IPasswordHasher, ICallerDetailProvider } from "../common/authentication";
import { errorFactories, type IDatabaseErrorHandler } from "../common/errors";
import type {
  AuthenticationVerifyUserNameAndPasswordRequestDto,
  AuthenticationChangePasswordRequestDto } from "@hc-management/shared/dtos";
import { OperationError } from "@hc-management/shared/errors";
import { errorMessages, getDisplayNameByKey } from "@hc-management/shared/localization";

type VerifyUserNameAndPasswordRequestDto = AuthenticationVerifyUserNameAndPasswordRequestDto;

export interface IAuthenticationService {
  verifyUserNameAndPasswordAsync(requestDto: VerifyUserNameAndPasswordRequestDto): Promise<void>;
  changePasswordAsync(requestDto: AuthenticationChangePasswordRequestDto): Promise<void>;
}

export class AuthenticationService implements IAuthenticationService {
  private readonly database: PrismaClient;
  private readonly passwordHasher: IPasswordHasher;
  private readonly callerDetailProvider: ICallerDetailProvider;
  private readonly databaseErrorHandler: IDatabaseErrorHandler;

  public constructor({ prisma, callerDetailProvider, passwordHasher, databaseErrorHandler }: IServiceContainer) {
    this.database = prisma;
    this.callerDetailProvider = callerDetailProvider;
    this.passwordHasher = passwordHasher;
    this.databaseErrorHandler = databaseErrorHandler;
  }

  public async verifyUserNameAndPasswordAsync(requestDto: VerifyUserNameAndPasswordRequestDto): Promise<void> {
    const user = await this.database.user.findUnique({
      where: { userName: requestDto.userName }
    });

    if (!user) {
      throw errorFactories.operationError.createNotFound("user");
    }

    if (!await this.passwordHasher.verifyPasswordAsync(requestDto.password, user.passwordHash)) {
      throw new OperationError({ "": errorMessages.incorrect(getDisplayNameByKey("password")) });
    }
  }

  public async changePasswordAsync(requestDto: AuthenticationChangePasswordRequestDto): Promise<void> {
    const caller = this.callerDetailProvider.getCallerDetail();
    const passwordHash = await this.passwordHasher.hashPasswordAsync(requestDto.newPassword);
    
    try {
      await this.database.user.update({
        data: { passwordHash },
        where: { id: caller.id }
      });
    } catch (error: any) {
      const handledResult = this.databaseErrorHandler.handle<User>(error);
      if (handledResult?.type === "RecordNotFound") {
        throw errorFactories.operationError.createNotFound("user");
      }

      throw error;
    }
  }
}

File: backend/Core/Features/Authentication/Services/AuthenticationService.cs
using FluentValidation;
using Microsoft.EntityFrameworkCore;
using HCManagement.Core.Common.Exceptions;
using HCManagement.Core.Common.Extensions;
using HCManagement.Core.Common.Localization;
using HCManagement.Core.Common.Security;
using HCManagement.Core.Features.Users;
using HCManagement.Core.Persistence.DbContext;
using HCManagement.Core.Persistence.Handlers;

namespace HCManagement.Core.Features.Authentication;

internal class AuthenticationService : IAuthenticationService
{
    #region Fields
    private readonly AppDbContext _context;
    private readonly IValidator<VerifyCredentialsRequestDto> _verifyCredentialsValidator;
    private readonly IValidator<ChangePasswordRequestDto> _changePasswordValidator;
    private readonly ICallerDetailProvider _callerDetailProvider;
    private readonly IPasswordHasher _passwordHasher;
    private readonly IDbExceptionHandler _exceptionHandler;
    #endregion
    
    #region Constructors
    public AuthenticationService(
        AppDbContext context,
        IValidator<VerifyCredentialsRequestDto> verifyCredentialsValidator,
        IValidator<ChangePasswordRequestDto> changePasswordValidator,
        ICallerDetailProvider callerDetailProvider,
        IPasswordHasher passwordHasher,
        IDbExceptionHandler exceptionHandler)
    {
        _context = context;
        _verifyCredentialsValidator = verifyCredentialsValidator;
        _changePasswordValidator = changePasswordValidator;
        _callerDetailProvider = callerDetailProvider;
        _passwordHasher = passwordHasher;
        _exceptionHandler = exceptionHandler;
    } 
    #endregion
    
    #region Methods
    public async Task VerifyCredentialsAsync(VerifyCredentialsRequestDto requestDto)
    {
        requestDto.TransformValues();
        _verifyCredentialsValidator.ValidateAndThrow(requestDto);
        
        string passwordHash = await _context.Users
            .Where(u => u.UserName == requestDto.UserName && u.DeletedDateTime == null)
            .Select(u => u.PasswordHash)
            .SingleOrDefaultAsync()
            ?? throw new OperationException(
               new object[] { nameof(requestDto.UserName) },
               ErrorMessages.NotFound.ReplaceResourceName(DisplayNames.User)
            );
        
        bool isPasswordCorrect = _passwordHasher.VerifyPassword(requestDto.Password, passwordHash);
        if (!isPasswordCorrect)
        {
            throw new OperationException(
                new object[] { nameof(requestDto.Password) },
                ErrorMessages.Incorrect.ReplacePropertyName(DisplayNames.Password)
            );
        }
    }

    public async Task ChangePasswordAsync(ChangePasswordRequestDto requestDto)
    {
        requestDto.TransformValues();
        _changePasswordValidator.ValidateAndThrow(requestDto);

        int callerId = _callerDetailProvider.GetId();
        User user = await _context.Users
            .Where(u => u.Id == callerId && u.DeletedDateTime == null)
            .SingleOrDefaultAsync()
            ?? throw new NotFoundException();
        
        if (!_passwordHasher.VerifyPassword(requestDto.CurrentPassword, user.PasswordHash))
        {
            throw new OperationException(
                new object[] { nameof(requestDto.CurrentPassword) },
                ErrorMessages.Incorrect.ReplacePropertyName(DisplayNames.CurrentPassword)
            );
        }
        
        string newPasswordHash = _passwordHasher.HashPassword(requestDto.NewPassword);
        user.PasswordHash = newPasswordHash;

        try
        {
            await _context.SaveChangesAsync();
        }
        catch (DbUpdateException exception)
        {
            DbExceptionHandledResult? handledResult = _exceptionHandler.Handle(exception);
            if (handledResult is null)
            {
                throw;
            }

            if (handledResult.IsConcurrencyConflict)
            {
                throw new ConcurrencyException();
            }

            throw;
        }
    }
    #endregion
}
